### DISCUSSION TITLE
Data into Database

### DISCUSSION THREAD

**Discussion by moses19850** [Hacksore/bluelinky#234] [Q&A] [answered]
Hi,
first thanks for all the efforts. I'm happy with this thing. I'm not a developer, far away from it, but I found a way to make use of it.

Now comes question: Does someone throw this data into a database to archive it?
If not, can you please at least explain me, how to extract single values out of it?
My goal is to use a simple html page on my synology to display the data.

Thanks a lot upfront.

this is the data which I'm using now:

`client.on("ready", async () => {
  const vehicle = client.getVehicle("********");
  const fullStatus = await vehicle.fullStatus({
	refresh: false,
    parsed: false
    });
  console.log('full status cached : ' + JSON.stringify(fullStatus, null, 2)); `

And as well:

`client.on("ready", async () => {
  const vehicle = client.getVehicle("****");
  const response = await vehicle.tripInfo({year: 2022, month: 12, day: 17});
  console.log('trips : ' + JSON.stringify(response, null, 2));
})`

Response for tripinfo:

trips : [
--
{
"dayRaw": "20221219",
"tripsCount": 2,
"distance": 4,
"durations": {
"drive": 11,
"idle": 1
},
"speed": {
"avg": 34,
"max": 68
},
"trips": [
{
"timeRaw": "195757",
"start": "2022-12-19T18:57:57.000Z",
"end": "2022-12-19T19:01:57.000Z",
"durations": {
"drive": 4,
"idle": 0
},
"speed": {
"avg": 39,
"max": 68
},
"distance": 2
},
{
"timeRaw": "080720",
"start": "2022-12-19T07:07:20.000Z",
"end": "2022-12-19T07:14:20.000Z",
"durations": {
"drive": 7,
"idle": 1
},
"speed": {
"avg": 29,
"max": 63
},
"distance": 2
}
]
}
]

Response for status:

full status cached   : {
--
"vehicleLocation": {
"coord": {
"lat": *****,
"lon": *******,
"alt": 3,
"type": 0
},
"head": 194,
"speed": {
"value": 0,
"unit": 0
},
"accuracy": {
"hdop": 10,
"pdop": 10
},
"time": "20221222141103"
},
"vehicleStatus": {
"airCtrlOn": false,
"engine": false,
"doorLock": true,
"doorOpen": {
"frontLeft": 0,
"frontRight": 0,
"backLeft": 0,
"backRight": 0
},
"trunkOpen": false,
"airTemp": {
"value": "01H",
"unit": 0,
"hvacTempType": 0
},
"defrost": false,
"lowFuelLight": false,
"acc": false,
"hoodOpen": false,
"transCond": false,
"steerWheelHeat": 0,
"sideBackWindowHeat": 0,
"dte": {
"value": 133,
"unit": 1
},
"tirePressureLamp": {
"tirePressureLampAll": 0
},
"battery": {
"batState": 0
},
"lampWireStatus": {
"stopLamp": {
"stopLampStatus": false,
"leftLamp": false,
"rightLamp": false
},
"headLamp": {
"headLampStatus": false,
"leftLowLamp": false,
"rightLowLamp": false,
"leftHighLamp": false,
"rightHighLamp": false
},
"turnSignalLamp": {
"turnSignalLampStatus": false,
"leftFrontLamp": false,
"rightFrontLamp": false,
"leftRearLamp": false,
"rightRearLamp": false
}
},
"windowOpen": {
"frontLeft": 0,
"frontRight": 0,
"backLeft": 0,
"backRight": 0
},
"fuelLevel": 10,
"washerFluidStatus": false,
"breakOilStatus": false,
"engineOilStatus": false,
"sleepModeCheck": true,
"time": "20221222141106"
},
"odometer": {
"value": 25256.2,
"unit": 1
}
}

**Comment by chubby1968** [comment]
Database: It depends on which kind you want to use. I'm not familiar with Synology, so I can't say what would be the easiest interface. In any case you would probably need to create tables for the data you want to store before creating commands to store the data.

Extracting a single value:
The response is an object so you can access a single value as follows (2 values are extracted in the example below):

```js
client.on("ready", async () => { 
  const vehicle = client.getVehicle("********"); 
  const fullStatus = await vehicle.fullStatus({ refresh: false, parsed: false });
  const time = fullStatus.vehicleLocation.time;
  const fuel = fullStatus.vehicleStatus.fuelLevel;
  console.log(`time: ${time}, fuel: ${fuel}`);
})
```

Also, if you just want to display the latest data on a HTML page, there's no need for a database.

**Reply by moses19850** [reply]
Thanks a lot. This is exactly what I want to achieve.
But I struggle now with "tripinfo". I don't know how to use the logic for this part.

First try:
`const response = await vehicle.tripInfo({year: 2022, month: 12, day: 22});
console.log(response)`

Result:
[
  {
    dayRaw: '20221222',
    tripsCount: 2,
    distance: 4,
    durations: { drive: 9, idle: 2 },
    speed: { avg: 37.5, max: 82 },
    trips: [ [Object], [Object] ]
  }
]

Next logical step for me:
`const response = await vehicle.tripInfo({year: 2022, month: 12, day: 22});
console.log(response.dayRaw);`

Result: undefined

Any  idea?

**Comment by chubby1968** [comment]
The result is inside an array, so it should be

```js
console.log(response[0].dayRaw);
```

**Comment by moses19850** [comment, marked as answer]
Hi,
thanks for the support so far.
With your help I'm able to fetch the data I need and convert into a proper structure. I'm also able to connect to my mariadb on synology.

My issue now: For fetchting the data I needed a loop and now I have no idea how to upload data from a loop into the database.

Any advice is appreciated.

Thanks

Code for fetching the data:
```
//const trpInfo = await vehicle.tripInfo({year: new Date().getFullYear(), month: new Date().getMonth()+1, day: new Date().getDate()});
  const trpInfo = await vehicle.tripInfo({year: 2022, month: 12, day: 4});

  if (trpInfo != '') 
    {
  
  //Erstelle Datum
  const tripdaydateraw = trpInfo[0].dayRaw.slice(4,6) + '/' + trpInfo[0].dayRaw.slice(6,8) + '/' + trpInfo[0].dayRaw.slice(0,4);
  const tripdaydate = trpInfo[0].dayRaw.slice(0,4) + '-' + trpInfo[0].dayRaw.slice(4,6) + '-' + trpInfo[0].dayRaw.slice(6,8);
  const tripdayname = new Date(tripdaydateraw).toLocaleDateString('de-de', {weekday: 'long'});
    //Erstelle Anzahl der Fahrten
  const tripdayCnt = trpInfo[0].tripsCount;
  //Erstelle Tagesdistanz
  const tripdayDistance = trpInfo[0].distance;
  //Erstelle Tagedurchschnittsgeschwindigkeit
  const tripdayavgSpeed = trpInfo[0].speed.avg;
  //Erstelle maximale Tagesgeschwindigkeit
  const tripdaymaxSpeed = trpInfo[0].speed.max;
  
  // Errechne Anzahl der Fahrten
  let i = 0, n = tripdayCnt - 1


  while(i <= n) {
      //Restliche Werte definieren

      //Erstelle Durchschnittsgeschwindigkeit Fahrt
      const tripavgSpeed = trpInfo[0].trips[i].speed.avg;
      //Erstelle maximal Fahrgeschwindigkeit
      const tripmaxSpeed = trpInfo[0].trips[i].speed.max;
      //Startzeit
      const strttimstamp = trpInfo[0].trips[i].start;
      //Enzeit
      const endtimstamp = trpInfo[0].trips[i].end;
      //Startdatum
      const strtdate = strttimstamp.toLocaleDateString();
      //toLocaleTimeString konvertiert GMT + 1, Startzeit
      const strtime = strttimstamp.toLocaleTimeString();
      //Enddatum
      const enddate = endtimstamp.toLocaleDateString();
      //toLocaleTimeString konvertiert GMT + 1, Endzeit
      const endtime = endtimstamp.toLocaleTimeString();

      // Fahrzeit von Minuten in das Format hh:mm umwandeln
      function toHoursAndMinutes(totalMinutes) {
        const minutes = totalMinutes % 60;
        const hours = Math.floor(totalMinutes / 60);
  
      return `${padTo2Digits(hours)}:${padTo2Digits(minutes)}`;
      }
  
      function padTo2Digits(num) {
        return num.toString().padStart(2, '0');
      }
  
      //Fahrzeit in Minuten
      const drvtime = toHoursAndMinutes(trpInfo[0].trips[i].durations.drive);

      //Distanz
      const drvdistance = trpInfo[0].trips[i].distance

      //Werte Ausgeben
      //const tripinfodata = tripdaydate + ';' + tripdayCnt + ';' + tripdayDistance + ';' + tripdayavgSpeed + ';' + tripdaymaxSpeed + ';' + tripavgSpeed + ';' + tripmaxSpeed + ';' + strtime + ';' + endtime + ';' + drvtime + ';' + drvdistance + ';' + tripdayname + ';;';
      const tripinfodata = tripdaydate + ',"' + tripdayCnt + '","' + tripdayDistance + '","' + tripdayavgSpeed + '","' + tripdaymaxSpeed + '","' + tripavgSpeed + '","' + tripmaxSpeed + '","' + strtime + '","' + endtime + '","' + drvtime + '","' + drvdistance + '","' + tripdayname + '";"";';
      //console.log(strtime)
      console.log(tripinfodata)
      i += 1;}};
});
```


Code for write data into mariadb:

```
// Use the MariaDB Node.js Connector
const mariadb = require('mariadb');
const { kill } = require('process');

// Create a connection pool
const pool = 
  mariadb.createPool({
    host: 'ipaddress', 
    port: port,
    user: 'db_user', 
    password: 'db_password',
    database: 'db'
});

pool.getConnection()
    .then(conn => {
    
      conn.query("INSERT INTO db_table(tripdayavgSpeed) VALUES('42.80')")
        .then((rows) => {
          console.log(rows); //[ {val: 1}, meta: ... ]
          pool.end();
		      conn.end();
        })
        .catch(err => {
          //handle error
          console.log(err); 
          pool.end();
          conn.end();
        })
        
    }).catch(err => {
      //not connected
      pool.end();
      conn.end();
    });
```

**Reply by chubby1968** [reply]
It's not really a Bluelinky question anymore, but you need to explore how to make SQL queries with the [mariaDB node connector](https://mariadb.com/kb/en/connector-nodejs-promise-api/#connectionquerysql-values-promise).

Here's a quick untested example (there might be timing issues since the queries are promise based):

```js
const mariadb = require('mariadb');
const pool = mariadb.createPool(...);

function insert_db( my_value ) {
    pool.getConnection()
      .then(conn => {
        conn.query(
	           "INSERT INTO db_table(tripdayavgSpeed) VALUES (?)", 
	           [ my_value ]
            )
            .then(...)
      })
     .catch(...)
}

for(let i = 0; i < 5; i++){
    insert_db(i);
}
```

A word of warning: Do take care that you don't make too many API calls to your car or you could drain your battery.

**Reply by moses19850** [reply]
thanks, but I struggle already on step before.
The result where I fetch the basic data looks like this:

"2022-12-04","3","237","75","134","51","108","17:40:34","17:52:34","12","8","Sonntag";"";
"2022-12-04","3","237","75","134","92","134","15:33:29","17:28:29","115","165","Sonntag";"";
"2022-12-04","3","237","75","134","82","133","09:44:15","10:40:15","56","64","Sonntag";"";

This is the result from the loop.
Now I'm fighting how to link this result to the use it for the SQL query.
I tried the "module.exports" but I'm too stupid to get it work. I was looking on google for it, without success

**Reply by chubby1968** [reply]
I would test a couple of variables before putting everything into the database. In the example below I've just used tripdaydate and tripdayCnt.

```js
// Define a function to do insert query
function insert_db( tripdaydate , tripdayCnt ) {
    pool.getConnection()
      .then(conn => {
        conn.query(
	           "INSERT INTO db_table(tripdayavgSpeed) VALUES (?,?)", 
	           [ tripdaydate , tripdayCnt ]
            )
            .then(...)
      })
     .catch(...)
}

// Whenever you want to insert actual values, you call the function e.g. in a loop
for(...){
   ...
   insert_db(tripdaydate,tripdayCnt);
}
```

And if you're struggling with the fact that you got some code in one file and some in another file you can export the function you need from the database script. E.g.

### "mariaDB" file
```js
export function insert_db( tripdaydate , tripdayCnt ) {
...
}
```

### "Fetching data" file

Then import in another file:

```js
import { insert_db } from "path/to/mariaDB.js";

for(...){
   ...
   insert_db(tripdaydate,tripdayCnt);
}
```

**Reply by moses19850** [reply]
Thanks for your help. I imported each single value into the database manually to make sure that everything works. The last step is really to put the values from the "fetching data" file into the "mariadb" file and this not work as I don't know how put the data into the "mariaDB" file.

**Reply by chubby1968** [reply]
Does it work if you keep everything in a single file?

E.g:

```js
// Use the MariaDB Node.js Connector
const mariadb = require('mariadb');

const trpInfo = await vehicle.tripInfo({ year: 2022, month: 12, day: 4 });

// Create a connection pool
const pool =
	mariadb.createPool({
		host: 'ipaddress',
		port: port,
		user: 'db_user',
		password: 'db_password',
		database: 'db'
	});

if (trpInfo != '') {

	//Erstelle Datum
	const tripdaydateraw = trpInfo[0].dayRaw.slice(4, 6) + '/' + trpInfo[0].dayRaw.slice(6, 8) + '/' + trpInfo[0].dayRaw.slice(0, 4);
	const tripdaydate = trpInfo[0].dayRaw.slice(0, 4) + '-' + trpInfo[0].dayRaw.slice(4, 6) + '-' + trpInfo[0].dayRaw.slice(6, 8);
	const tripdayname = new Date(tripdaydateraw).toLocaleDateString('de-de', { weekday: 'long' });
	//Erstelle Anzahl der Fahrten
	const tripdayCnt = trpInfo[0].tripsCount;
	//Erstelle Tagesdistanz
	const tripdayDistance = trpInfo[0].distance;
	//Erstelle Tagedurchschnittsgeschwindigkeit
	const tripdayavgSpeed = trpInfo[0].speed.avg;
	//Erstelle maximale Tagesgeschwindigkeit
	const tripdaymaxSpeed = trpInfo[0].speed.max;

	// Errechne Anzahl der Fahrten
	let i = 0, n = tripdayCnt - 1;


	while (i <= n) {
		//Restliche Werte definieren

		//Erstelle Durchschnittsgeschwindigkeit Fahrt
		const tripavgSpeed = trpInfo[0].trips[i].speed.avg;
		//Erstelle maximal Fahrgeschwindigkeit
		const tripmaxSpeed = trpInfo[0].trips[i].speed.max;
		//Startzeit
		const strttimstamp = trpInfo[0].trips[i].start;
		//Enzeit
		const endtimstamp = trpInfo[0].trips[i].end;
		//Startdatum
		const strtdate = strttimstamp.toLocaleDateString();
		//toLocaleTimeString konvertiert GMT + 1, Startzeit
		const strtime = strttimstamp.toLocaleTimeString();
		//Enddatum
		const enddate = endtimstamp.toLocaleDateString();
		//toLocaleTimeString konvertiert GMT + 1, Endzeit
		const endtime = endtimstamp.toLocaleTimeString();

		// Fahrzeit von Minuten in das Format hh:mm umwandeln
		function toHoursAndMinutes(totalMinutes) {
			const minutes = totalMinutes % 60;
			const hours = Math.floor(totalMinutes / 60);

			return `${padTo2Digits(hours)}:${padTo2Digits(minutes)}`;
		}

		function padTo2Digits(num) {
			return num.toString().padStart(2, '0');
		}

		//Fahrzeit in Minuten
		const drvtime = toHoursAndMinutes(trpInfo[0].trips[i].durations.drive);

		//Distanz
		const drvdistance = trpInfo[0].trips[i].distance;

		//Werte Ausgeben
		const tripinfodata = tripdaydate + ',"' + tripdayCnt + '","' + tripdayDistance + '","' + tripdayavgSpeed + '","' + tripdaymaxSpeed + '","' + tripavgSpeed + '","' + tripmaxSpeed + '","' + strtime + '","' + endtime + '","' + drvtime + '","' + drvdistance + '","' + tripdayname + '";"";';
		//console.log(strtime)
		pool.getConnection()
			.then(conn => {

				conn.query("INSERT INTO db_table(tripdayavgSpeed) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?)", [tripdaydate, tripdayCnt, tripdayDistance, tripdayavgSpeed, tripdaymaxSpeed, tripavgSpeed, tripmaxSpeed, strtime, endtime, drvtime, drvdistance, tripdayname])
					.then((rows) => {
						console.log(rows); //[ {val: 1}, meta: ... ]
						pool.end();
						conn.end();
					})
					.catch(err => {
						//handle error
						console.log(err);
						pool.end();
						conn.end();
					});

			}).catch(() => {
				//not connected
				pool.end();
				conn.end();
			});
		console.log(tripinfodata);
		i += 1;
	}
};
```

**Reply by moses19850** [reply]
You made my day. Thanks a lot. After really small tweaks it works. If you need the correct code, please let me know.
And related to the battery topic: My plan is to run this once per day around midnight, and anyhow, this data is fetched from the server and not from the car itself, as far as I understood from code.

Thanks again for your support and your time.
